File: 23-merge-k-sorted-lists/merge-k-sorted-lists.js
var mergeKLists = function(lists) {
	// use MinPriorityQueue class from the datastructures-js library that is available in the LeetCode runtime
    const minHeap = new MinPriorityQueue({ priority: x => x.val });
    
    // push head of each list into heap
    for (const head of lists) {
        if (head) {
            minHeap.enqueue(head);
        }
    }
    
	// create head as dummy node
    let head = new ListNode();
    let tail = head;
    
    while (!minHeap.isEmpty()) {
        const node = minHeap.dequeue().element;
        tail.next = node;
        tail = tail.next;
        if (node.next) {
            minHeap.enqueue(node.next);
        }
    }
    return head.next;
};

File: 23-merge-k-sorted-lists/merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists:
            return None

        heap = []

        for i, head in enumerate(lists):
            if head:
                heapq.heappush(heap, (head.val, i, head))

        dummy = ListNode()
        curr = dummy

        while heap:
            value, list_idx, node = heapq.heappop(heap)

            curr.next = node
            curr = curr.next

            if node.next:
                heapq.heappush(heap, (node.next.val, list_idx, node.next))

        return dummy.next